core/src/main/scala/doric/types/LiteralSparkType.scala

```diff
@@ -0,0 +1,136 @@
+package doric
+package types
+
+import scala.reflect.ClassTag
+
+import java.sql.{Date, Timestamp}
+import java.time.{Instant, LocalDate}
+
+trait LiteralSparkType[T] {
+  self =>
+  type OriginalSparkType
+
+  val classTag: ClassTag[OriginalSparkType]
+
+  val literalTo: T => OriginalSparkType
+
+  def customType[O](
+      f: O => T
+  )(implicit
+      ost: SparkType[O]
+  ): LiteralSparkType.Custom[O, self.OriginalSparkType] =
+    new LiteralSparkType[O]() {
+      override type OriginalSparkType = self.OriginalSparkType
+      val classTag: ClassTag[self.OriginalSparkType] = self.classTag
+      override val literalTo: O => OriginalSparkType = f andThen self.literalTo
+    }
+}
+
+object LiteralSparkType {
+
+  type Primitive[T] = LiteralSparkType[T] {
+    type OriginalSparkType = T
+  }
+
+  type Custom[T, O] = LiteralSparkType[T] {
+    type OriginalSparkType = O
+  }
+
+  @inline def apply[T](implicit
+      litc: LiteralSparkType[T]
+  ): Custom[T, litc.OriginalSparkType] =
+    litc
+
+  @inline private def createPrimitive[T: ClassTag](implicit
+      spark: SparkType.Primitive[T]
+  ): Primitive[T] = new LiteralSparkType[T] {
+    override type OriginalSparkType = spark.OriginalSparkType
+    val classTag: ClassTag[T]                      = implicitly[ClassTag[T]]
+    override val literalTo: T => OriginalSparkType = identity
+  }
+
+  implicit val fromNull: Primitive[Null] = createPrimitive[Null]
```

Comment:
I do not know how to do it, but this seems a bit odd as we have to maintain implicit values for `SparkType` and `LiteralSparkType`... Every spark type should have a LiteralSparkType as well, don't you think?

Reply:
Some cant, like Row, because it can't be used as literal.